code/code/misc/discord.cc

```diff
@@ -16,97 +19,137 @@ namespace po = boost::program_options;
 
 // available discord channels
 // channels are configured in lib/discord.cfg
-// new channels must be added in that file, here, in discord.h, and in po::options_descriptions below
+// new channels must be added in that file, here, in discord.h, and in
+// po::options_descriptions below
 sstring Discord::CHANNEL_DEATHS;
 sstring Discord::CHANNEL_SYS;
 sstring Discord::CHANNEL_ACHIEVEMENT;
 
 // threshold level for discord mob kill notifications
 int Discord::ACHIEVEMENT_THRESHOLD;
 
+// for thread management
+bool Discord::stop_thread;
+std::thread Discord::messenger_thread;
+std::queue<std::pair<sstring, sstring>> Discord::message_queue;
+std::mutex Discord::queue_mutex;
+
 // read the configuration
 bool Discord::doConfig() {
-    using std::string;
-    string configFile = "discord.cfg";
-
-    std::string empty_string = "";
-
-    po::options_description configOnly("Configuration File Only");
-    configOnly.add_options()
-        ("deaths_webhook",po::value<string>(&CHANNEL_DEATHS)->default_value(empty_string),"see discord.h")
-        ("sys_webhook",po::value<string>(&CHANNEL_SYS)->default_value(empty_string),"see discord.h")
-        ("achieve_webhook",po::value<string>(&CHANNEL_ACHIEVEMENT)->default_value(empty_string),"see discord.h")
-        ("achieve_threshold",po::value<int>(&ACHIEVEMENT_THRESHOLD)->default_value(80),"see discord.h");
-
-    po::options_description config_options;
-    config_options.add(configOnly);
-
-    po::variables_map vm;
-    po::notify(vm);
-    std::ifstream ifs(configFile.c_str());
-
-    if (!ifs.is_open()) {
-        vlogf(LOG_MISC, format("Discord webhooks: Failed to open config file '%s'") % configFile);
-        vlogf(LOG_MISC, format("Discord webhooks: ensure config is located in: '%s'") % std::filesystem::current_path());
-        return false;
-    }
-
-    po::store(parse_config_file(ifs, config_options), vm);
-    po::notify(vm);
-
-    return true;
+  using std::string;
+  // worker thread setup
+  stop_thread = false;
```

Comment:
Can we provide that default value where we declare or define this variable?
Or do we expect doConfig to be called multiple times per boot?

Reply:
I'm looking at adding a simple sleep to the thread:
```
    if (message_queue.empty() && !stop_thread) {
      std::this_thread::sleep_for(std::chrono::milliseconds(1000)); // sleep for a second
      continue;
    }
    ```

Reply:
A sleep incurs a context switch, unlike an event. When waiting for an event, the kernel removes the thread completely from scheduling.